src/app/loadout/mod-utils.ts

```diff
@@ -299,6 +310,82 @@ export function getCheapestModAssignments(
   return { itemModAssignments: mergedResults, unassignedMods };
 }
 
+/**
+ * This orders the results of the mod assignment algorithm for a given item and its mods.
+ *
+ * Basically
+ * - It will find all the relevant sockets
+ * - Assign mods to each socket or assign the default plug if there are no mods for the socket
+ * - Then order the results so that we never go over the energy limit if we were to assign them one
+ *   by one.
+ */
+function createOrderedAssignmentResults(
+  defs: D2ManifestDefinitions | undefined,
+  item: DimItem,
+  bucketSpecificAssignments: PluggableInventoryItemDefinition[],
+  bucketIndependentAssignments: PluggableInventoryItemDefinition[]
+) {
+  const results: {
+    socketIndex: number;
+    mod?: PluggableInventoryItemDefinition;
+    // This will be negative if we are recovering used energy back by swapping in a cheaper mod
+    energyChange: number;
+  }[] = [];
+
+  const mods = [...bucketIndependentAssignments, ...bucketSpecificAssignments];
+
+  const modIndexes =
+    item.sockets?.categories.find(
+      (category) => category.category.hash === SocketCategoryHashes.ArmorMods
+    )?.socketIndexes || [];
+  const modSockets = getSocketsByIndexes(item.sockets!, modIndexes);
+
+  for (const mod of mods) {
+    // Check to see if its already socketed somewhere
+    let sortedSocketIndex = modSockets.findIndex(
+      (socket) => socket.plugged?.plugDef.hash === mod.hash
+    );
+    if (sortedSocketIndex < 0) {
+      // Otherwise find the first socket it can fit into
+      sortedSocketIndex = modSockets.findIndex(
+        (socket) => socket.plugged?.plugDef.plug.plugCategoryHash === mod.plug.plugCategoryHash
+      );
+    }
+
+    // This should always be the case, if its not then something is seriously wrong
+    if (sortedSocketIndex > 0) {
+      const pluggedCost =
+        modSockets[sortedSocketIndex].plugged?.plugDef.plug.energyCost?.energyCost || 0;
+      const modCost = mod.plug.energyCost?.energyCost || 0;
+      const energyChange = modCost - pluggedCost;
+      results.push({ socketIndex: modSockets[sortedSocketIndex].socketIndex, mod, energyChange });
+      modSockets.splice(sortedSocketIndex, 1);
```

Comment:
Note we always add the assignment, we don't check if its already socketed. We do this for the visual display in the UI. We skip that specific case right before we dispatch the assignment action